open-telemetry/opentelemetry-js
# fix(sdk-trace-base): Export processed spans while exporter failed

## Which problem is this PR solving?

While the exporter deals with a batch of spans, new spans may come in and wait to be exported. As previously implemented, a successful export would notice these waiting spans, triggering a renewed timer check, but not so for an unsuccessful export. 

The result was that, prior to this commit, a failing export may end up in a situation where no further spans will be exported. This is due to the behaviour of `_addToBuffer` when the queue is full: Imagine an export which fails after a long timeout (because of, for instance, network troubles). While the connection waits to be timed out, the span queue fills up. Once completely full, no new calls to recheck the timer will be done. On its own, this behaviour is fine. When combined with the patched bug, this leads to a rather confusing case where the exporter never tries exporting.

## Short description of the changes

This PR makes sure these spans are acknowledged and handled, even when exporting fails.

@pichlermarc you've mentioned working on refactoring the exporter; here's an interesting edge case.

## Type of change

Please delete options that are not relevant.


## How Has This Been Tested?

- Unit test
- Functional testing against a real-world setup, before and after the patch:
  - Start the instrumented application once with a working collector, see that everything flows
  - Start the application again, this time without a collector listening on the configured port; wait for for a few failed exports, then run the collector again, seeing whether spans arrive

## Checklist:
